src/lib/statistics/chunk_statistics/histograms/abstract_histogram.hpp

```diff
@@ -0,0 +1,214 @@
+ * The possible maximum length of the prefix depends on the number of supported characters.
+ * It must not overflow the uint64_t data type used to represent strings as numbers.
+ * The formula used to verify the prefix length is:
+ * string_prefix_length < std::log(std::numeric_limits<uint64_t>::max()) / std::log(supported_characters.length() + 1)
+ */
+template <typename T>
+class AbstractHistogram : public AbstractFilter {
+ public:
+  AbstractHistogram();
+  AbstractHistogram(const std::string& supported_characters, const uint32_t string_prefix_length);
+  virtual ~AbstractHistogram() = default;
+
+  /**
+   * Strings are internally transformed to a number, such that a bin can have a numerical width.
+   * This transformation is based on uint64_t.
+   */
+  using HistogramWidthType = std::conditional_t<std::is_same_v<T, std::string>, uint64_t, T>;
+
+  virtual HistogramType histogram_type() const = 0;
+
+  /**
+   * Returns a string with detailed information about the histogram, including the edges of the individual bins.
+   */
+  std::string description() const;
+
+  /**
+   * Returns the estimated selectivity given a predicate type and its parameter(s).
+   * It will always be between 0 and 1.
+   */
+  float estimate_selectivity(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                             const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Returns the estimated cardinality given a predicate type and its parameter(s).
+   * It will always be between 0 and total_count().
+   * This method is specialized for strings to handle predicates uniquely applicable to string columns.
+   */
+  float estimate_cardinality(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                             const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Returns whether a given predicate type and its parameter(s) can be pruned.
+   * This method is specialized for strings to handle predicates uniquely applicable to string columns.
+   */
+  bool can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                 const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const override;
+
+  /**
+   * Returns the lower bound (minimum value) of the histogram.
+   * This is equal to the smallest value in the segment.
+   */
+  T min() const;
+
+  /**
+   * Returns the upper bound (maximum value) of the histogram.
+   * This is equal to the largest value in the segment.
+   */
+  T max() const;
+
+  /**
+   * Returns the number of bins actually present in the histogram.
+   * This number can differ from the number of bins requested when creating a histogram.
+   */
+  virtual BinID bin_count() const = 0;
+
+  /**
+   * Returns the number of values represented in the histogram.
+   * This is equal to the length of the segment during creation, without null values.
```

Comment:
What does "during creation" mean here?

Reply:
Is this better?

```c++
/**
   * Returns the number of values represented in the histogram.
   * This is equal to the number of rows in the segment during the generation of the bins for the histogram,
   * without null values.
   */
```

Or is the comment is more confusing than helpful and we should just go with the first part?

Reply:
Ah, so you are thinking about the case where a histogram is created on a chunk that is not yet finished?

Reply:
Yes, or if values in the segment are updated to `NULL`, then the histogram does not know about that. Also, currently, MVCC information is not considered at all.